src/gurt/shm_dict.c

```diff
@@ -0,0 +1,430 @@
+/**
+ * (C) Copyright 2024 Intel Corporation.
+ * (C) Copyright 2025 Hewlett Packard Enterprise Development LP
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+#include <errno.h>
+
+#include "shm_internal.h"
+#include <gurt/shm_dict.h>
+
+/* the address of shared memory region */
+extern struct d_shm_hdr *d_shm_head;
+
+int
+shm_ht_create(const char name[], int bits, int n_lock, struct d_shm_ht_head **ht_head)
+{
+	int                   i;
+	struct d_shm_ht_head *ht_head_loc;
+	long int             *off_next;
+	long int              offset;
+	d_shm_mutex_t        *p_locks;
+	int                   len_name;
+	int                   n_bucket;
+
+	*ht_head = NULL;
+	len_name = strnlen(name, MAX_HT_NAME_LEN);
+	if (len_name >= MAX_HT_NAME_LEN) {
+		DS_ERROR(EINVAL, "hash table name is longer than %d bytes", MAX_HT_NAME_LEN - 1);
+		return EINVAL;
+	}
+
+	/* force bucket size be power of 2 */
+	n_bucket = 1 << bits;
+	shm_mutex_lock(&(d_shm_head->ht_lock), NULL);
+
+	/* loop over existing hash tables to check whether it exists or not */
+	if (d_shm_head->off_ht_head != INVALID_OFFSET) {
+		offset = d_shm_head->off_ht_head;
+		while (offset != INVALID_OFFSET) {
+			ht_head_loc = (struct d_shm_ht_head *)((char *)d_shm_head + offset);
+			if ((strncmp(name, ht_head_loc->ht_name, MAX_HT_NAME_LEN) == 0) &&
+			    (ht_head_loc->n_bucket == n_bucket) &&
+			    (ht_head_loc->n_lock == n_lock)) {
+				/* found existing hash table */
+				*ht_head = ht_head_loc;
+				break;
+			}
+			if (ht_head_loc->next == INVALID_OFFSET) {
+				/* reaching the end of the link list of existing hash tables */
+				*ht_head = NULL;
+				break;
+			}
+			offset = ht_head_loc->next;
+		}
+	}
+
+	if (*ht_head) {
+		/* found existing hash table */
+		shm_mutex_unlock(&(d_shm_head->ht_lock));
+		return 0;
+	}
+
+	/* This hash table does not exist, then create it. */
+	*ht_head = shm_alloc(sizeof(struct d_shm_ht_head) + (sizeof(d_shm_mutex_t) * n_lock) +
+			     (sizeof(long int) * n_bucket));
+	if (*ht_head == NULL)
+		return ENOMEM;
+	ht_head_loc = *ht_head;
+
+	memcpy(ht_head_loc->ht_name, name, len_name + 1);
+	ht_head_loc->n_bucket = n_bucket;
```

Comment:
At least we should check if `n_bucket` is greater or equal than `n_lock`.
We could also check if `n_bucket % n_lock == 0` as it will ensure an even number of buckets per lock and also not need to use float to find the lock index.

Reply:
This is fixed as suggested. Thank you!